### DISCUSSION TITLE
request: log `topics` type could be narrowed

### DISCUSSION THREAD

**Discussion by typedarray** [wevm/viem#163] [Idea / Feature Request] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.1.6

### Current Behavior

Event logs returned by viem have a topics property [typed](https://github.com/wagmi-dev/viem/blob/main/src/types/log.ts#L59) like this:

```ts
type TopicFilter = (null | string | string[])[]

type Log = {
  /** List of order-dependent topics */
  topics: TopicFilter
}
```

### Expected Behavior

While the `topics` property of the filter argument to RPC methods like `eth_getLogs` and `eth_newFilter` can contain elements that are an array of strings or null, I'm fairly certain that the `topics` property for any given `Log` object  returned by these RPC methods will only ever contain strings ([Solidity docs reference](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html#events)).

```ts
type Topics = string[]

type Log = {
  /** List of order-dependent topics */
  topics: Topics
}
```

It may be possible to narrow the type further like this:

```ts
type Hex = `0x${string}`

type Topics = [Hex, Hex | undefined, Hex | undefined, Hex | undefined]

```

but I'm not certain that the first topic must be defined (can an anonymous event have 0 indexed arguments?).

**Comment by ryansmith3136** [comment, marked as answer]
Yellow Paper section 4.3.1:

A log entry, $O$, is a tuple of the logger's address, $O_a$, a possibly empty series of 32-byte log topics, $O_{\mathbf{t}}$ and some number of bytes of data, $O_{\mathbf{d}}$:

$$O \equiv (O_{\mathrm{a}}, ({O_{\mathbf{t}}}_0, {O_{\mathbf{t}}}_1, ...), O_{\mathbf{d}})$$

Topics can be empty. For example: [this](https://etherscan.io/tx/0x8732bf3e88a17addd06c29885e1bd1a0d62272f77b8a0e60b0e28a632d7b7a57#eventlog) transaction contains logs with empty topics.

**Comment by typedarray** [comment]
Thanks! So, the best type is probably:
```ts
type Topics = Hex[];
```

Although it is guaranteed that the length of the array will be <= 4, and that could be represented as a union of tuples:
```ts
type Topics = [] | [Hex] | [Hex, Hex] | [Hex, Hex, Hex] | [Hex, Hex, Hex, Hex];
```

In any case, the current type of `(null | string | string[])[]` is incorrect because AFAIK topic elements will never be `null` and will never be `string[]`.
